### ISSUE TITLE
Add type hints for files in <repo root>/arelle/

### ISSUE THREAD

**Issue by ggravlingen** [Arelle/Arelle#314]
We should add type hints to the library. To make coordination easier, I suggest using this issue to mention what files we are/will be working on and tick the ones already covered. As a follow-up, we could work on the files in the sub-folders.

- [ ] __init__.py (empty file)
- [ ] _pkg_meta.py
- [ ] Cntlr.py
- [ ] CntlrCmdLine.py
- [ ] CntlrComServer.py
- [ ] CntlrProfiler.py
- [ ] CntlrQuickBooks.py
- [ ] CntlrWebMain.py
- [ ] CntlrWinMain.py
- [ ] CntlrWinTooltip.py
- [ ] DialogAbout.py
- [ ] DialogArcroleGroup.py
- [ ] DialogFind.py
- [ ] DialogFormulaParameters.py
- [ ] DialogLanguage.py
- [ ] DialogNewFactItem.py
- [ ] DialogOpenArchive.py
- [ ] DialogOpenTaxonomyPackage.py
- [ ] DialogPackageManager.py
- [ ] DialogPluginManager.py
- [ ] DialogRssWatch.py
- [ ] DialogURL.py
- [ ] DialogUserPassword.py
- [ ] DisclosureSystem.py
- [ ] FileSource.py
- [ ] FormulaConsisAsser.py
- [ ] FormulaEvaluator.py
- [ ] FunctionCustom.py
- [ ] FunctionFn.py
- [ ] FunctionIxt.py
- [ ] FunctionUtil.py
- [ ] FunctionXfi.py
- [ ] FunctionXs.py
- [ ] HashUtil.py
- [ ] HtmlUtil.py
- [ ] InstanceAspectsEvaluator.py
- [ ] LeiUtil.py
- [ ] Locale.py
- [ ] LocalViewer.py
- [ ] ModelDocument.py
- [ ] ModelDtsObject.py
- [ ] ModelFormulaObject.py
- [ ] ModelInstanceObject.py
- [ ] ModelManager.py
- [ ] ModelObject.py
- [ ] ModelObjectFactory.py
- [ ] ModelRelationshipSet.py
- [ ] ModelRenderingObject.py
- [ ] ModelRssItem.py
- [ ] ModelRssObject.py
- [ ] ModelTestcaseObject.py
- [ ] ModelValue.py
- [ ] ModelVersObject.py
- [ ] ModelVersReport.py
- [ ] ModelXbrl.py
- [ ] PackageManager.py
- [ ] PluginManager.py
- [ ] PrototypeDtsObject.py
- [ ] PrototypeInstanceObject.py
- [ ] PythonUtil.py
- [ ] RenderingEvaluator.py
- [ ] RenderingResolver.py
- [ ] TableStructure.py
- [ ] TkTableWrapper.py
- [ ] typing.py
- [ ] UITkTable.py
- [ ] UiUtil.py
- [ ] Updater.py
- [ ] UrlUtil.py
- [ ] Validate.py
- [ ] ValidateFilingText.py
- [ ] ValidateFormula.py
- [ ] ValidateInfoset.py
- [ ] ValidateUtr.py
- [ ] ValidateVersReport.py
- [ ] ValidateXbrl.py
- [ ] ValidateXbrlCalcs.py
- [ ] ValidateXbrlDimensions.py
- [ ] ValidateXbrlDTS.py
- [ ] Version.py
- [ ] ViewFile.py
- [ ] ViewFileConcepts.py
- [ ] ViewFileDTS.py
- [ ] ViewFileFactList.py
- [ ] ViewFileFactTable.py
- [ ] ViewFileFormulae.py
- [ ] ViewFileRelationshipSet.py
- [ ] ViewFileRenderedGrid.py
- [ ] ViewFileRoleTypes.py
- [ ] ViewFileRssFeed.py
- [ ] ViewFileTests.py
- [ ] ViewUtil.py
- [ ] ViewUtilFormulae.py
- [ ] ViewWinConcepts.py
- [ ] ViewWinDiffs.py
- [ ] ViewWinDTS.py
- [ ] ViewWinFactGrid.py
- [ ] ViewWinFactList.py
- [ ] ViewWinFactTable.py
- [ ] ViewWinFormulae.py
- [ ] ViewWinGrid.py
- [ ] ViewWinList.py
- [ ] ViewWinPane.py
- [ ] ViewWinProperties.py
- [ ] ViewWinRelationshipSet.py
- [ ] ViewWinRenderedGrid.py
- [ ] ViewWinRoleTypes.py
- [ ] ViewWinRssFeed.py
- [ ] ViewWinTests.py
- [ ] ViewWinTkTable.py
- [ ] ViewWinTree.py
- [ ] ViewWinTupleGrid.py
- [ ] ViewWinVersReport.py
- [ ] ViewWinXml.py
- [ ] WatchRss.py
- [ ] WebCache.py
- [ ] XbrlConst.py
- [ ] XbrlUtil.py
- [ ] XhtmlValidate.py
- [ ] XmlUtil.py
- [ ] XmlValidate.py
- [ ] XmlValidateParticles.py
- [ ] XmlValidateSchema.py
- [ ] XPathContext.py
- [ ] XPathParser.py

**Comment by selgamal** [comment]
Agreed, this should be allocated gradually, I mean someone should line up the files, and whom ever want to work on this tasj should take the next file in line.

**Comment by ggravlingen** [comment]
I'm planning on working on `ModelXbrl` next. I have a sense we will get a lot of help by adding type hints to some of the core files first.

**Comment by selgamal** [comment]
Yes thats what I am saying, there should be some logical order for the file being worked on.

**Comment by ggravlingen** [comment]
Maybe the Workiva team has an idea of a good route? Like a Machete order for Arelle.

**Comment by selgamal** [comment]
> I'm planning on working on `ModelXbrl` next. I have a sense we will get a lot of help by adding type hints to some of the core files first.

If you are targeting `ModelXbrl`,  in my opinion the route that makes sense is Filesource -> ModelDocument -> ModelXbrl

**Comment by sagesmith-wf** [comment]
FYI @selgamal @ggravlingen I think @austinmatherne-wk is going to look into organizing the typing project in some fashion. In the meantime I suggest working with him to facilitate grabbing files to type

**Comment by ggravlingen** [comment]
> FYI @selgamal @ggravlingen I think @austinmatherne-wk is going to look into organizing the typing project in some fashion. In the meantime I suggest working with him to facilitate grabbing files to type

Noted, thanks for the heads up!

**Comment by ggravlingen** [comment]
I've pondered over this and suggest a strategy where we only touch one file at at a time (and add `# type: ignore[no-untyped-call]`) to functions defined in other files. I believe that will be a lot easier to review, rather than touching files everywhere in the project. Would love to hear your thoughts on this @austinmatherne-wk and @sagesmith-wf.

**Comment by sagesmith-wf** [comment]
> I've pondered over this and suggest a strategy where we only touch one file at at a time (and add # type: ignore[no-untyped-call]) to functions defined in other files. I believe that will be a lot easier to review, rather than touching files everywhere in the project. When the other files are typed, mypy will raise error: Unused "type: ignore" comment and we can just clean up the ignore comments.

I think that is an excellent idea. I would push us to do a pass and ignore all untyped files as a whole with the comment and type them 1 by 1 though

**Comment by selgamal** [comment]
Just an idea, how about generating stubs and then annotating them, I personally prefer to add type hints to the code, but stubs seem to be easier to manage and does not bloat existing code, and it seems to be the preferred way to go for existing projects.
[Stubs]( https://mypy.readthedocs.io/en/stable/stubs.html) also see [stubgen](https://mypy.readthedocs.io/en/stable/stubgen.html) for generating stubs for existing code.

**Comment by selgamal** [comment]
@austinmatherne-wk  I am planning to work on [XmlUtil.py](https://github.com/Arelle/Arelle/blob/master/arelle/XmlUtil.py) or whatever you recommend to do next, please let me know.

**Comment by ggravlingen** [comment]
My five cents is that the XMLUtil file would be good to cover. It'll save us a lot of type: ignore to add type hints to that file.

**Comment by austinmatherne-wk** [comment]
@selgamal go for it!

**Comment by selgamal** [comment]
@ggravlingen @sagesmith-wf  I will be working on XbrlConst.py and later today if I can I will start with XbrlUtil.py, if anyone working on those please let me know.

**Comment by ggravlingen** [comment]
👍 I am not working on any of those.

**Comment by ggravlingen** [comment]
I've started looking into `XmlValidate`

**Comment by anjm2** [comment]
Hi, I'd like to work on adding type hints to arelle/CntlrProfiler.py and removing it from the mypy ignore_errors override. Is this module still available to work on?

**Comment by austinmatherne-wk** [comment]
Hi @anjm2 ,

Thanks for reminding me of this issue. The tracking list was out of date. CntlrProfiler is also no longer used. I opened #2601 to delete it.

**Comment by anjm2** [comment]
Hi @austinmatherne-wk 
Thanks for the update and for cleaning up the tracking list! I’m still interested in contributing to the type-hinting effort. Could you recommend one of the remaining smaller modules that would be suitable for a first contribution?

**Comment by austinmatherne-wk** [comment]
The `View*` modules are probably the simplest of what's left. Many of the remaining modules are complicated. I wouldn't suggest starting any of the rendering or formula modules unless you already understand Table Linkbase and XBRL formula very well. I'd also recommend reading our [CONTRIBUTING.md](https://github.com/Arelle/Arelle/blob/master/CONTRIBUTING.md) doc. Good luck!
